components/script/dom/decompressionstream.rs

```diff
@@ -0,0 +1,303 @@
+            Decompressor::Gzip(gz_decoder) => gz_decoder.write_all(buf),
+        }
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        match self {
+            Decompressor::Deflate(zlib_decoder) => zlib_decoder.flush(),
+            Decompressor::DeflateRaw(deflate_decoder) => deflate_decoder.flush(),
+            Decompressor::Gzip(gz_decoder) => gz_decoder.flush(),
+        }
+    }
+
+    fn try_finish(&mut self) -> io::Result<()> {
+        match self {
+            Decompressor::Deflate(zlib_decoder) => zlib_decoder.try_finish(),
+            Decompressor::DeflateRaw(deflate_decoder) => deflate_decoder.try_finish(),
+            Decompressor::Gzip(gz_decoder) => gz_decoder.try_finish(),
+        }
+    }
+}
+
+/// <https://compression.spec.whatwg.org/#decompressionstream>
+#[dom_struct]
+pub(crate) struct DecompressionStream {
+    reflector_: Reflector,
+
+    /// <https://streams.spec.whatwg.org/#generictransformstream>
+    transform: Dom<TransformStream>,
+
+    /// <https://compression.spec.whatwg.org/#decompressionstream-format>
+    format: CompressionFormat,
+
+    // <https://compression.spec.whatwg.org/#decompressionstream-context>
+    #[ignore_malloc_size_of = "defined in flate2"]
+    #[no_trace]
+    context: RefCell<Decompressor>,
+}
+
+impl DecompressionStream {
+    fn new_inherited(
+        transform: &TransformStream,
+        format: CompressionFormat,
+    ) -> DecompressionStream {
+        DecompressionStream {
+            reflector_: Reflector::new(),
+            transform: Dom::from_ref(transform),
+            format,
+            context: RefCell::new(Decompressor::new(format)),
+        }
+    }
+
+    fn new_with_proto(
+        global: &GlobalScope,
+        proto: Option<SafeHandleObject>,
+        transform: &TransformStream,
+        format: CompressionFormat,
+        can_gc: CanGc,
+    ) -> DomRoot<DecompressionStream> {
+        reflect_dom_object_with_proto(
+            Box::new(DecompressionStream::new_inherited(transform, format)),
+            global,
+            proto,
+            can_gc,
+        )
+    }
+}
+
+impl DecompressionStreamMethods<crate::DomTypeHolder> for DecompressionStream {
+    /// <https://compression.spec.whatwg.org/#dom-decompressionstream-decompressionstream>
+    fn Constructor(
+        global: &GlobalScope,
+        proto: Option<SafeHandleObject>,
+        can_gc: CanGc,
+        format: CompressionFormat,
+    ) -> Fallible<DomRoot<DecompressionStream>> {
+        // Step 1. If format is unsupported in DecompressionStream, then throw a TypeError.
+        // NOTE: All of "deflate", "deflate_raw" and "gzip" are supported.
+
+        // Step 2. Set this’s format to format.
+        // Step 5. Set this’s transform to a new TransformStream.
+        let transform = TransformStream::new_with_proto(global, None, can_gc);
+        let this = DecompressionStream::new_with_proto(global, proto, &transform, format, can_gc);
```

Comment:
I would avoid using `this` and `that` in naming, `Set this’s` means to set current DecompressionStream's format